packages/fxa-auth-server/lib/payments/subscription-reminders.ts

```diff
@@ -0,0 +1,216 @@
+    private reminderLength: number = DEFAULT_REMINDER_LENGTH,
+    db: any,
+    mailer: any
+  ) {
+    this.db = db;
+    this.mailer = mailer;
+    this.stripeHelper = Container.get(StripeHelper);
+  }
+
+  /**
+   * For all possible plan.intervals, determine if the plan is sufficiently
+   * long based on planLength.
+   */
+  private isEligiblePlan(plan: Plan): boolean {
+    for (let [interval, checkFn] of Object.entries(eligiblePlansByInterval)) {
+      if (plan.interval === interval && checkFn(plan, this.planLength)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private async getEligiblePlans(): Promise<Plan[]> {
+    const allPlans = await this.stripeHelper.allPlans();
+    return allPlans.filter((plan) => this.isEligiblePlan(plan));
+  }
+
+  /**
+   * Returns a window of time in seconds [startTimeS, endTimeS)
+   * that is exactly one day, reminderLength days from now in UTC.
+   */
+  private getStartAndEndTimes(reminderLengthMs: number): TimeSpanInS {
+    const reminderDay = new Date(Date.now() + reminderLengthMs);
+    // Get hour 0, minute 0, second 0 for today's date
+    const startingTimestamp = new Date(
+      Date.UTC(
+        reminderDay.getUTCFullYear(),
+        reminderDay.getUTCMonth(),
+        reminderDay.getUTCDate(),
+        0,
+        0,
+        0
+      )
+    );
+    // Get hour 0, minute, 0, second 0 for one day from today's date
+    const endingTimestamp = new Date(startingTimestamp.getTime() + MS_IN_A_DAY);
+
+    const startTimeS = Math.floor(startingTimestamp.getTime() / 1000);
+    const endTimeS = Math.floor(endingTimestamp.getTime() / 1000);
+    return {
+      startTimeS,
+      endTimeS,
+    };
+  }
+
+  private async alreadySentEmail(
+    uid: string,
+    currentPeriodStartMs: number,
+    emailParams: SentEmailParams
+  ) {
+    const emailRecord = await SentEmail.findLatestSentEmailByType(
+      uid,
+      EMAIL_TYPE,
+      emailParams
+    );
+    // This could be the first email for a given subscription or a subsequent one.
+    return !!emailRecord && emailRecord.sentAt > currentPeriodStartMs;
+  }
+
+  private async updateSentEmail(uid: string, emailParams: SentEmailParams) {
+    await SentEmail.createSentEmail(uid, EMAIL_TYPE, emailParams);
+  }
+
+  /**
+   * Send out a renewal reminder email if we haven't already sent one.
+   */
+  async sendSubscriptionRenewalReminderEmail(
+    subscription: Stripe.Subscription
+  ): Promise<Boolean> {
+    const { customer } = subscription;
+    // @ts-ignore
+    const { email } = customer;
+    // @ts-ignore
+    const uid = customer.metadata.userid;
+    if (!uid || !email) {
+      // Deleted customer
+      return false;
+    }
+    const emailParams = { subscriptionId: subscription.id };
+    if (
+      await this.alreadySentEmail(
+        uid,
+        Math.floor(subscription.current_period_start * 1000),
+        emailParams
+      )
+    ) {
+      return false;
```

Comment:
Should we throw here? I am envisioning we may run this script 1-2x/day, which means we could see some dupe subscriptions that have already been processed, and that wouldn’t be an error. So I’m just returning.

Reply:
Sound reasoning. I don't think we need to throw if we already sent an email.